sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala

```diff
@@ -157,6 +157,38 @@ final class OffsetWindowFunctionFrame(
   override def currentUpperBound(): Int = throw new UnsupportedOperationException()
 }
 
+class FixedOffsetWindowFunctionFrame(
+    target: InternalRow,
+    ordinal: Int,
+    expressions: Array[OffsetWindowFunction],
+    inputSchema: Seq[Attribute],
+    newMutableProjection: (Seq[Expression], Seq[Attribute]) => MutableProjection,
+    offset: Int)
+  extends OffsetWindowFunctionFrame(
+    target, ordinal, expressions, inputSchema, newMutableProjection, offset) {
+
+  override def prepare(rows: ExternalAppendOnlyUnsafeRowArray): Unit = {
+    input = rows
+    inputIndex = 0
+  }
+
+  override def write(index: Int, current: InternalRow): Unit = {
```

Comment:
NTH_VALUE should return the same value for all rows in the window partition right? So why are you doing so much heavy lifting here? Everything can be computed in prepare.

If you think about it, then this could also be treated as an unbounded window frame. You could even move this into the `UnboundedWindowFunctionFrame` if you add the update logic to the NTH_VALUE aggregate.

Reply:
Because the behavior of `NTH_VALUE/FIRST_VALUE/LAST_VALUE` more similar to `LEAD/LAG`, so I make `FixedOffsetWindowFunctionFrame`  extends `OffsetWindowFunctionFrame`. `OffsetWindowFunctionFrame` only to fetch projects but `UnboundedWindowFunctionFrame` will `update` and `execute` expressions.
If you think so, I can abstract a base class for `FixedOffsetWindowFunctionFrame` and `OffsetWindowFunctionFrame`.

Reply:
I have optimized the code so that the skip operator runs once in `prepare`.